<a href="https://colab.research.google.com/github/Valiev-Koyiljon/Statistics/blob/main/TS_and_ES_Pearson_Correlation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pearson Correlation
a.k.a. Pearson’s r, Pearson Product Moment Correlation, bivariate correlation

*By P. Stikker*<br>
https://PeterStatistics.com<br>
https://www.youtube.com/stikpet<br>

## Introduction

The most commonly used measure to test if a linear relation exists between two scale variables, is the Pearson Correlation Coefficient (or in it's full name the Pearson product-moment correlation coefficient) (Pearson, 1896).

Pearson Correlation varies between -1 and +1. If it is -1 there is a perfect negative lineair relationship, if it is 0 there is no lineair relationship and at +1 there is a perfect positive lineair relationship.

A positive relation means that if one variable goes up, the other also goes up (for example number of ice cream sold versus temperature), a negative relation indicates if one goes down, the other goes up (for example number of winter jackets sold versus temperature).

Unfortunately there is no formal way to determine if a particular value is high or low, and the rules of thumb floating around on the internet vary quite a lot, often depending on the field (e.g. biology, medicine, business, etc.). For example the same rule of thumb sizes from Rea and Parker (2014):

|\|r\|| Interpretation|
|-------|---------------|
|0.00 < 0.10| Negligible|
|0.10 < 0.20 |Weak|
|0.20 < 0.40| Moderate|
|0.40 < 0.60| Relatively strong|
|0.60 < 0.80| Strong|
|0.80 <= 1.00| Very strong|


We can also test if Pearson Correlation might be significantly different from 0.

## Example

To show an example, I'll load some data as a pandas dataframe. So I'll need the '<a href="https://pandas.pydata.org">pandas</a>' library:

In [ ]:
#!pip install pandas
import pandas as pd

And then load the example data using the <a href="https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html">'read_csv'</a>. 

In [ ]:
myDf = pd.read_csv('../../Data/csv/EmployeeData.csv')
myDf.head()

id  gender         bdate  educ    jobcat   salary  salbegin  jobtime  \
0  1.0    Male  1.165415e+10  15.0   Manager  57000.0   27000.0     98.0   
1  2.0    Male  1.185296e+10  16.0  Clerical  40200.0   18750.0     98.0   
2  3.0  Female  1.094334e+10  12.0  Clerical  21450.0   12000.0     98.0   
3  4.0  Female  1.150252e+10   8.0  Clerical  21900.0   13200.0     98.0   
4  5.0    Male  1.174936e+10  15.0  Clerical  45000.0   21000.0     98.0   

  prevexp minority  
0   144.0       No  
1    36.0       No  
2   381.0       No  
3   190.0       No  
4   138.0       No

For the example I'll use the *salbegin* (the beginning salary) and the *salary* (the current salary). We can create a separate dataframe from just those two, and remove any missing values from it.

In [ ]:
newDf = myDf[['salbegin', 'salary']].dropna()
newDf.head()

salbegin   salary
0   27000.0  57000.0
1   18750.0  40200.0
2   12000.0  21450.0
3   13200.0  21900.0
4   21000.0  45000.0

Pandas has its own correlation coefficient function 'corr'. We can simply add it as a function to the new dataframe:

In [ ]:
newDf.corr()

salbegin    salary
salbegin  1.000000  0.880117
salary    0.880117  1.000000

A correlation coefficient of 0.88. It is positive so if the beginning salary is high then also the current salary will be usually higher. 
The relation could be classified as 'very strong', since it fits in the 0.80 <= 1.00 category.

We can get the same result using numpy. Of course we first import numpy:

In [ ]:
# !pip install numpy
import numpy as np

Then convert our two fields to numpy arrays.

In [ ]:
npArr1 = np.array(myDf['salary'])
npArr2 = np.array(myDf['salbegin'])

Then we can use numpy's 'corrcoef' function:

In [ ]:
np.corrcoef(npArr1, npArr2)

array([[1.        , 0.88011747],
       [0.88011747, 1.        ]])

The same result as with pandas. 

Neither pandas nor numpy will actually show the signficance (p-value), but there are more packages available. One of them is for example scipy.stats:

In [ ]:
# !pip install scipy.stats
from scipy.stats import pearsonr

Now we can use the 'pearsonr' function:

In [ ]:
pearsonr(myDf['salary'], myDf['salbegin'])

(0.8801174655999495, 8.203222401982817e-155)

The first value is the correlation coefficient, and the second the p-value (significance). 

In the example the significance of this test is .000. This is the chance of finding a correlation coefficient of .880 or even higher in a sample, if in the population it would be 0 (no association). This is such a low chance, that we can say that in the population the correlation coefficient will be indeed different from zero, and conclude that there is a significant linear association between the two variables.

Also pingouin shows this:

In [ ]:
# !pip install pingouin
from pingouin import corr

In [ ]:
corr(myDf['salary'], myDf['salbegin'])

n         r        CI95%        r2   adj_r2          p-val  \
pearson  474  0.880117  [0.86, 0.9]  0.774607  0.77365  8.203222e-155   

               BF10  power  
pearson  6.638e+150    1.0

Or researchpy

In [ ]:
# !pip install researchpy
from researchpy.correlation import corr_case

In [ ]:
corr_case(myDf[['salary', 'salbegin']])

(  Pearson correlation test using list-wise deletion
 0                     Total observations used = 474,
           salary salbegin
 salary         1   0.8801
 salbegin  0.8801        1,
           salary salbegin
 salary    0.0000   0.0000
 salbegin  0.0000   0.0000)

Again the same results. There probably more packages out there, but this should give some indication.

In the appendix I'll go over the formulas avoiding packages almost entirely (only for the t-distribution its used to get the p-value).

## References

Pearson, K. (1896). Mathematical contributions to the theory of evolution. III. Regression, heredity, and panmixia. *Philosophical Transactions of the Royal Society of London. (A.), 1896*, 253–318.

Rea, L. M., & Parker, R. A. (2014). *Designing and conducting survey research: a comprehensive guide*. San Francisco: Jossey-Bass Publishers.

## Appendix: The hard way

To avoid packages as much as possible, we will need to go over the formulas. First convert the two variables scores into a Python native format: a list.

In [ ]:
list1 = list(newDf['salary'])
list2 = list(newDf['salbegin'])

The formula for the Spearman rho can be written as:

\begin{equation*}
r_s = \frac{\sum_{i=1}^n \left( \left( x_i-\bar{x} \right) \times \left(y_i-\bar{y}\right)\right)}{\sqrt{SS_{x}\times SS_{y}}}
\end{equation*}

We'll disect this formula in parts.

First there is $\bar{x}$ and $\bar{y}$. These are the means of each of the two lists.

\begin{equation*}
\bar{x} = \frac{\sum_{i=1}^n x_i}{n}, \bar{y} = \frac{\sum_{i=1}^n y_i}{n}
\end{equation*}

In these formulas $n$ is the number of pairs, $x_i$ is the i-th score in the first variable, and $y_i$ the i-th score in the second variable.
With Pandas 'len' function we can easily determine the number of pairs, and with Pandas 'sum' function we can sum the ranks of each variable.

In [ ]:
n = len(list1)
mean1 = sum(list1) / n
mean2 = sum(list2) / n

n, mean1, mean2

(474, 34419.56751054852, 17016.086497890294)

Next is that numerator:

\begin{equation*}
\sum_{i=1}^n \left( \left( x_i-\bar{x} \right) \times \left(y_i-\bar{y}\right)\right)
\end{equation*}

We actually have everything needed here, so simply plug in the formula:

In [ ]:
num = 0
for i in range(n):
    num = num + (list1[i] - mean1) * (list2[i] - mean2)
    
num

55948605047.73208

The denominator looked like:

\begin{equation*}
\sqrt{SS_{x}\times SS_{y}}
\end{equation*}

Here the $SS_{x}$ is the sum of squares, defined with:

\begin{equation*}
SS_{x} = \sum_{i=1}^n \left( x_i - \bar{x} \right)^2
\end{equation*}

and similar for $SS_{y}$:
\begin{equation*}
SS_{y} = \sum_{i=1}^n \left( y_i - \bar{y} \right)^2
\end{equation*}

We have all we need for these, so can immediately code them:

In [ ]:
SS1 = 0
for i in list1:
    SS1 = SS1 + (i - mean1)**2

SS2 = 0
for i in list2:
    SS2 = SS2 + (i - mean2)**2

SS1, SS2

(137916495436.33975, 29300904965.45357)

Finally we can simply fill out the formula now for Spearman rho:

In [ ]:
pearsonCorr = num / (SS1 * SS2)**0.5
pearsonCorr

0.880117465599948

To get the p-value, we can convert the coefficient to a t-value using:

\begin{equation*}
t = r \times \sqrt{\frac{n - 2}{1 - r^2}}
\end{equation*}

In [ ]:
tval = pearsonCorr * ((n - 2) / (1 - pearsonCorr**2))**0.5
tval

40.27552303306331

To find the corresponding p-value to this t-value we do need a library that has the t-distribution, like scipy.stats:

In [ ]:
# !pip install scipy.stats
from scipy.stats import t

We also need the degrees of freedom, which is defined as:

\begin{equation*}
df = n - 2
\end{equation*}


In [ ]:
df = n - 2
df

472

Finally for the p-value we can now use:

In [ ]:
t.sf(tval, df)*2

8.203222402012955e-155